### PR TITLE
Improved error handling

### PR DESCRIPTION
**Pull request by neworderofjamie**
This is based on something I added to Pinus Rigida (with a lot of input from @mossblaser) - It makes figuring out what crashed a lot easier. Again, some git magic is probably required to rebase it onto something useful/cherry pick it into master.

### CODE REVIEW

**Comment by mossblaser** [comment]
> This is based on something I added to Pinus Rigida (with a lot of input from @mossblaser)

No idea what that was but I'll happily take credit ;) :P

**Comment by neworderofjamie** [comment]
"How should I handle errors Jonathan? HOW? HOW?"

File: nengo_spinnaker/simulator.py [hunk 1]
```diff
@@ -255,17 +254,9 @@ def _run_steps(self, steps):
...
-                x, y = self.netlist.placements[vertex]
-                p = self.netlist.allocations[vertex][Cores].start
-                status = self.controller.get_processor_status(p, x, y)
-                if status.cpu_state is not AppState.sync0:
-                    print("Core ({}, {}, {}) in state {!s}".format(
-                        x, y, p, status.cpu_state))
-            raise Exception("Unexpected core failures.")
+        # Wait for cores to exit or re-enter sync0
+        self._wait_for_transition(AppState.run, [AppState.exit, AppState.sync0],
+                                  len(self.netlist.vertices))
```
Comment (mundya) [inline nengo_spinnaker/simulator.py:259]: Why wait for `AppState.exit`?

**Comment by neworderofjamie** [comment]
Well the executables call spin1_exit if they've run for the full simulation time - wouldn't that leave them in exit?

**Comment by mundya** [comment]
> Well the executables call spin1_exit if they've run for the full simulation time - wouldn't that leave them in exit?

No, it leaves them in `sync0` because of loops like https://github.com/project-rig/nengo_spinnaker/blob/master/spinnaker_components/filter/filter.c#L373-L383

**Comment by neworderofjamie** [comment]
Could you just stop the timer at this point instead then? https://github.com/project-rig/nengo_spinnaker/blob/master/spinnaker_components/filter/filter.c#L276-L279

**Comment by mundya** [comment]
Sorry, somehow didn't notice the update to this.

Thanks @neworderofjamie!

**Pull request merged**